### PR TITLE
executor: let information_schema be the first database in ShowDatabases

### PR DESCRIPTION
**Pull request by parastorli**
### What problem does this PR solve? 
show databases;
INFORMATION_SCHEMA is not first database

### What is changed and how it works?
Add one function to replace INFORMATION_SCHEMA to the first database if the first database is not INFORMATION_SCHEMA

### Check List
Tests 

 - Unit test
 - Integration test

### CODE REVIEW

**Comment by shenli** [comment]
@parastorli Thanks! Please make the description clear and add a test case.

**Comment by parastorli** [comment]
@shenli, Thabnks.
make the description clear, is about commit or comment? please further point 
please give me a link about how to add test and run the unit test to the tidb

**Comment by shenli** [comment]
I mean the checklist part. Please remove the items that are not applicable.

**Comment by parastorli** [comment]
I have changed the checklist, ok?

**Comment by parastorli** [comment]
@shenli Make the checklist clear and add the test.

File: executor/show_test.go [hunk 1]
```diff
@@ -462,6 +462,31 @@ func (s *testSuite) TestShowVisibility(c *C) {
...
+	tk.MustExec("create database AAAA")
+	tk.MustExec("create database bbbb")
+	tk.MustExec(`grant select on AAAA.* to 'show'@'%'`)
+	tk.MustExec(`grant select on bbbb.* to 'show'@'%'`)
+	tk.MustExec(`flush privileges`)
+
+	tk1 := testkit.NewTestKit(c, s.store)
+	se, err := session.CreateSession4Test(s.store)
+	c.Assert(err, IsNil)
+	c.Assert(se.Auth(&auth.UserIdentity{Username: "show", Hostname: "%"}, nil, nil), IsTrue)
```
Comment (ngaut) [inline executor/show_test.go]: s/"/```/

File: executor/show.go [hunk 2]
```diff
@@ -166,11 +166,28 @@ func (e *ShowExec) fetchShowEngines() error {
...
+func replaceInformationFirst(dbs []string) {
+	if strings.EqualFold(dbs[0], "INFORMATION_SCHEMA") {
+		return
+	}
+
+	first := dbs[0]
+	for i := 1; i < len(dbs); i++ {
+		if strings.EqualFold(dbs[i], "INFORMATION_SCHEMA") {
+			dbs[i] = first
+			dbs[0] = "INFORMATION_SCHEMA"
```
Comment (zz-jason) [inline executor/show.go]: how about:
```go
copy(dbs[1:i], dbs[0:i-1])
dbs[0] = "INFORMATION_SCHEMA"
```

Thus we have moved the "INFORMATION_SCHEMA" to the first element and the remained elements are sorted in the origin ascending order.

File: executor/show.go [hunk 3]
```diff
@@ -166,11 +166,28 @@ func (e *ShowExec) fetchShowEngines() error {
 	return nil
 }
 
+// let information_schema be the first database
+func replaceInformationFirst(dbs []string) {
```
Comment (zz-jason) [inline executor/show.go]: how about s/replaceInformationFirst/moveInfoSchemaToFront/?

Comment (parastorli) [reply executor/show.go]: good name.

File: executor/show.go (hunk 2 above)
Comment (parastorli) [reply executor/show.go]: good, Thanks!

Comment (parastorli) [reply executor/show.go]: Another question:
how to run just one file or one test case for the function,  like TestShowVisibility in the show_test.go to replace make dev for all test files.

File: executor/show_test.go (hunk 1 above)
Comment (parastorli) [reply executor/show_test.go]: Thanks!

File: executor/show.go (hunk 3 above)
Comment (parastorli) [reply executor/show.go]: Another question:
how to run just one file or one test case for the function,  like TestShowVisibility in the show_test.go to replace make dev for all test files.

Comment (zz-jason) [reply executor/show.go]: If the test resides in the xx directory:
```sh
cd xx
go test -check.f TestShowVisibility
```

Comment (parastorli) [reply executor/show.go]: Thanks!

**Comment by parastorli** [comment]
@ngaut 
@zz-jason 
@shenli 
please review the new pr.

File: executor/show.go [hunk 4]
```diff
@@ -166,11 +166,30 @@ func (e *ShowExec) fetchShowEngines() error {
...
 
+// move information_schema to the first, and the other are sorted in the origin ascending order.
+func moveInfoSchemaToFront(dbs []string) {
+	if strings.EqualFold(dbs[0], "INFORMATION_SCHEMA") {
+		return
+	}
+
+	for i := 1; i < len(dbs); i++ {
+		if strings.EqualFold(dbs[i], "INFORMATION_SCHEMA") {
+			tmp := dbs[i-1]
```
Comment (zz-jason) [inline executor/show.go]: this line can be removed.

File: executor/show.go [hunk 5]
```diff
@@ -166,11 +166,30 @@ func (e *ShowExec) fetchShowEngines() error {
...
+			// the copy num is i-1, dbs[i-1] isn't copied to dbs[i]
+			copy(dbs[1:i], dbs[0:i-1])
+			dbs[i] = tmp
```
Comment (zz-jason) [inline executor/show.go]: ditto

File: executor/show.go (hunk 2 above)
Comment (zz-jason) [reply executor/show.go]: Oh, I think it should be `copy(dbs[1:i+1], dbs[0:i])`

File: executor/show.go (hunk 5 above)
Comment (parastorli) [reply executor/show.go]: if just retain copy,
`	for i := 1; i < len(dbs); i++ {
		if strings.EqualFold(dbs[i], "INFORMATION_SCHEMA") {
			//tmp := dbs[i-1]
			// the copy num is i-1, dbs[i-1] isn't copied to dbs[i]
			copy(dbs[1:i], dbs[0:i-1])
			//dbs[i] = tmp
			dbs[0] = "INFORMATION_SCHEMA"
			return
		}
	}`
read the test case, run the test case
`cd executor && go test -check.f TestMoveInfoSchemaToFront`
 the error will be:
`
show_test.go:483:
    tk1.MustQuery("show databases").Check(testkit.Rows("INFORMATION_SCHEMA", "AAAA", "BBBB"))
/Users/andy/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:55:
    res.c.Assert(resBuff.String(), check.Equals, needBuff.String(), res.comment)
... obtained string = "" +
...     "[INFORMATION_SCHEMA]\n" +
...     "[AAAA]\n" +
...     "[INFORMATION_SCHEMA]\n"
... expected string = "" +
...     "[INFORMATION_SCHEMA]\n" +
...     "[AAAA]\n" +
...     "[BBBB]\n"
... sql:show databases, args:[]

`

File: executor/show.go (hunk 4 above)
Comment (parastorli) [reply executor/show.go]: I want to replace to copy(dbs[1:i+1], dbs[0:i]) before. but the i+1 in the dbs[1:i+1] is not in the work Interval. so I changed  to the code in the file.

**Comment by parastorli** [comment]
I want to replace to copy(dbs[1:i+1], dbs[0:i]) before. but the i+1 in the dbs[1:i+1] is not in the work Interval, it is dangerous. so I changed  to the code in the file.

**Comment by parastorli** [comment]
@zz-jason, please review the new pr.

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL

**Comment by parastorli** [comment]
@parastorli PTAL

**Comment by parastorli** [comment]
@parastorli ok

File: executor/show_test.go [hunk 6]
```diff
@@ -462,6 +462,31 @@ func (s *testSuite) TestShowVisibility(c *C) {
 	tk.MustExec("drop database showdatabase")
 }
 
+// ShowDatabases show the first database is INFORMATION_SCHEMA
+func (s *testSuite) TestMoveInfoSchemaToFront(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustQuery("show databases").Check(testkit.Rows("INFORMATION_SCHEMA"))
+	tk.MustExec(`create user 'show'@'%'`)
```
Comment (XuHuaiyu) [inline executor/show_test.go]: We can not use `show databases` without this in test?

File: executor/show_test.go [hunk 7]
```diff
@@ -462,6 +462,31 @@ func (s *testSuite) TestShowVisibility(c *C) {
 	tk.MustExec("drop database showdatabase")
 }
 
+// ShowDatabases show the first database is INFORMATION_SCHEMA
```
Comment (XuHuaiyu) [inline executor/show_test.go]: 1. Remove this comment.
2. This is an integration test but not a unit test, rename this func as TestShowDatabases.
3. Add a unit test for `moveInfoSchemaToFront`.

File: executor/show.go [hunk 8]
```diff
@@ -166,11 +166,28 @@ func (e *ShowExec) fetchShowEngines() error {
 	return nil
 }
 
+// move information_schema to the first, and the other are sorted in the origin ascending order.
```
Comment (XuHuaiyu) [inline executor/show.go]: s/ move/ moveInfoSchemaToFront moves
s/ the other/ the others

File: executor/show.go [hunk 9]
```diff
@@ -166,11 +166,28 @@ func (e *ShowExec) fetchShowEngines() error {
...
+func moveInfoSchemaToFront(dbs []string) {
+	if strings.EqualFold(dbs[0], "INFORMATION_SCHEMA") {
+		return
+	}
+
+	for i := 1; i < len(dbs); i++ {
+		if strings.EqualFold(dbs[i], "INFORMATION_SCHEMA") {
+			// the copy num is i, dbs[i] isn't copied to dbs[i+1]
```
Comment (XuHuaiyu) [inline executor/show.go]: remove this line

File: executor/show.go [hunk 10]
```diff
@@ -166,11 +166,28 @@ func (e *ShowExec) fetchShowEngines() error {
 	return nil
 }
 
+// move information_schema to the first, and the other are sorted in the origin ascending order.
+func moveInfoSchemaToFront(dbs []string) {
+	if strings.EqualFold(dbs[0], "INFORMATION_SCHEMA") {
+		return
+	}
+
+	for i := 1; i < len(dbs); i++ {
```
Comment (winoros) [inline executor/show.go]: Could you try binary search `SearchStrings` here? It won't increase the coding complexity.

File: executor/show_test.go (hunk 7 above)
Comment (parastorli) [reply executor/show_test.go]: Thanks. is the unit test about moveInfoSchemaToFront  in the file of the show_test.go?

File: executor/show.go (hunk 8 above)
Comment (parastorli) [reply executor/show.go]: Good!

File: executor/show.go (hunk 10 above)
Comment (parastorli) [reply executor/show.go]: Thanks, could have a try.

File: executor/show_test.go (hunk 7 above)
Comment (XuHuaiyu) [reply executor/show_test.go]: yep

Comment (parastorli) [reply executor/show_test.go]: but the package is executor_test, not executor, I need create one file, or can I use other files?

**Comment by parastorli** [comment]
I find in the dir of the executor, they are all integration tests，can you give me one example about the unit test in the dir?

File: executor/show_test.go (hunk 7 above)
Comment (XuHuaiyu) [reply executor/show_test.go]: ok, we can not call moveInfoSchemaToFront in package executor_test,
you can put the unit test in execuotr/pkg_test.go

File: executor/show_test.go (hunk 6 above)
Comment (XuHuaiyu) [reply executor/show_test.go]: Why do we need to create a new user?

Comment (parastorli) [reply executor/show_test.go]: Oh, I have deleted the cmd.

**Comment by parastorli** [comment]
@XuHuaiyu 
@winoros
PTAL

**Comment by parastorli** [comment]
@zz-jason 
jenkins-ci-tidb/integration-compatibility-test 
jenkins-ci-tidb/integration-ddl-test 
are failed, but are not related to the pr，why?

**Comment by XuHuaiyu** [comment]
Hi, @parastorli 
It has nothing to do with this PR, we are working on it.

**Pull request merged**